storage/paths/index.go

```diff
@@ -116,6 +283,156 @@ func (i *Index) StorageList(ctx context.Context) (map[storiface.ID][]storiface.D
 	return out, nil
 }
 
+func union(a, b []string) []string {
+	m := make(map[string]bool)
+
+	for _, elem := range a {
+		m[elem] = true
+	}
+
+	for _, elem := range b {
+		if _, ok := m[elem]; !ok {
+			a = append(a, elem)
+		}
+	}
+	return a
+}
+
+func splitString(str string) []string {
+	if str == "" {
+		return []string{}
+	}
+	return strings.Split(str, ",")
+}
+
+func (dbi *DBIndex) StorageAttach(ctx context.Context, si storiface.StorageInfo, st fsutil.FsStat) error {
+	var allow, deny = make([]string, 0, len(si.AllowTypes)), make([]string, 0, len(si.DenyTypes))
+
+	if _, hasAlert := dbi.pathAlerts[si.ID]; dbi.alerting != nil && !hasAlert {
+		dbi.pathAlerts[si.ID] = dbi.alerting.AddAlertType("sector-index", "pathconf-"+string(si.ID))
+	}
+
+	var hasConfigIssues bool
+
+	for id, typ := range si.AllowTypes {
+		_, err := storiface.TypeFromString(typ)
+		if err != nil {
+			//No need to hard-fail here, just warn the user
+			//(note that even with all-invalid entries we'll deny all types, so nothing unexpected should enter the path)
+			hasConfigIssues = true
+
+			if dbi.alerting != nil {
+				dbi.alerting.Raise(dbi.pathAlerts[si.ID], map[string]interface{}{
+					"message":   "bad path type in AllowTypes",
+					"path":      string(si.ID),
+					"idx":       id,
+					"path_type": typ,
+					"error":     err.Error(),
+				})
+			}
+
+			continue
+		}
+		allow = append(allow, typ)
+	}
+	for id, typ := range si.DenyTypes {
+		_, err := storiface.TypeFromString(typ)
+		if err != nil {
+			//No need to hard-fail here, just warn the user
+			hasConfigIssues = true
+
+			if dbi.alerting != nil {
+				dbi.alerting.Raise(dbi.pathAlerts[si.ID], map[string]interface{}{
+					"message":   "bad path type in DenyTypes",
+					"path":      string(si.ID),
+					"idx":       id,
+					"path_type": typ,
+					"error":     err.Error(),
+				})
+			}
+
+			continue
+		}
+		deny = append(deny, typ)
+	}
+	si.AllowTypes = allow
+	si.DenyTypes = deny
+
+	if dbi.alerting != nil && !hasConfigIssues && dbi.alerting.IsRaised(dbi.pathAlerts[si.ID]) {
+		dbi.alerting.Resolve(dbi.pathAlerts[si.ID], map[string]string{
+			"message": "path config is now correct",
+		})
+	}
+
+	for _, u := range si.URLs {
+		if _, err := url.Parse(u); err != nil {
+			return xerrors.Errorf("failed to parse url %s: %w", si.URLs, err)
+		}
+	}
+
+	// TODO: make below part of a single transaction
```

Comment:
Could probably be done with a single `insert into .. on conflict do update set ...` query. See e.g. https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-upsert/

Reply:
This is what gpt4 came up with when I asked it to rewrite this with a single query:

```go
var urls sql.NullString
var storageId sql.NullString

// Prepare the URL list for the new storage location
currUrls := strings.Join(union(strings.Split(urls.String, ","), si.URLs), ",")

// Prepare the INSERT ... ON CONFLICT ... DO UPDATE query
_, err = dbi.harmonyDB.Exec(ctx,
	`INSERT INTO StorageLocation (
		storage_id, 
		urls, 
		weight, 
		max_storage, 
		can_seal, 
		can_store, 
		groups, 
		allow_to, 
		allow_types, 
		deny_types,
		capacity,
		available,
		fs_available,
		reserved,
		used,
		created_at
	) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16)
	ON CONFLICT (storage_id) 
	DO UPDATE SET 
		urls = EXCLUDED.urls,
		weight = EXCLUDED.weight,
		max_storage = EXCLUDED.max_storage,
		can_seal = EXCLUDED.can_seal,
		can_store = EXCLUDED.can_store,
		groups = EXCLUDED.groups,
		allow_to = EXCLUDED.allow_to,
		allow_types = EXCLUDED.allow_types,
		deny_types = EXCLUDED.deny_types`,
	si.ID,
	currUrls,
	si.Weight,
	si.MaxStorage,
	si.CanSeal,
	si.CanStore,
	strings.Join(si.Groups, ","),
	strings.Join(si.AllowTo, ","),
	strings.Join(si.AllowTypes, ","),
	strings.Join(si.DenyTypes, ","),
	st.Capacity,
	st.Available,
	st.FSAvailable,
	st.Reserved,
	st.Used,
	time.Now())

if err != nil {
	log.Errorf("StorageAttach insert/update fails: ", err)
	return xerrors.Errorf("StorageAttach insert/update fails: ", err)
}
return nil
```

Reply:
Ah this is neat Postgres feature! Most other langs make you use temp tables which is a pain

Reply:
Unfortunately, you still need the first select since you need to create the updated url list so still need a transaction but this does consolidate the upsert operation